icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -483,37 +545,50 @@ class Parser {
                     status = kUnitIdentifierSyntaxError;
                     return;
             }
-        }
 
-        // We ran out of tokens before finding a complete single unit.
-        status = kUnitIdentifierSyntaxError;
+            previ = fIndex;
+            token = nextToken(status);
+            if (U_FAILURE(status)) {
+                // We have run out of (valid) tokens before finding a complete
+                // single unit.
+                return;
+            }
+        }
     }
 
+    /// @param result is modified, not overridden. Caller must pass in a
+    /// default-constructed (empty) MeasureUnitImpl instance.
     void parseImpl(MeasureUnitImpl& result, UErrorCode& status) {
         if (U_FAILURE(status)) {
             return;
         }
+        if (fSource.empty()) {
+            // The dimenionless unit: nothing to parse. leave result as is.
+            return;
+        }
         int32_t unitNum = 0;
+        if (!hasNext()) {
+            // We don't support the empty string.
+            status = kUnitIdentifierSyntaxError;
+            return;
+        }
         while (hasNext()) {
-            bool sawAnd;
+            bool sawAnd = false;
             SingleUnitImpl singleUnit;
             nextSingleUnit(singleUnit, sawAnd, status);
             if (U_FAILURE(status)) {
                 return;
             }
-            if (singleUnit.index == 0) {
-                continue;
-            }
+            U_ASSERT(!singleUnit.isDimensionless());
             bool added = result.append(singleUnit, status);
             if (sawAnd && !added) {
                 // Two similar units are not allowed in a mixed unit
                 status = kUnitIdentifierSyntaxError;
                 return;
             }
             if ((++unitNum) >= 2) {
-                UMeasureUnitComplexity complexity = sawAnd
-                    ? UMEASURE_UNIT_MIXED
-                    : UMEASURE_UNIT_COMPOUND;
+                UMeasureUnitComplexity complexity =
+                    sawAnd ? UMEASURE_UNIT_MIXED : UMEASURE_UNIT_COMPOUND;
                 if (unitNum == 2) {
                     U_ASSERT(result.complexity == UMEASURE_UNIT_SINGLE);
                     result.complexity = complexity;
```

Comment:
You catch mixed and/per sooner than this code did before.  Is it still possible for `result.complexity != complexity` on the following line?  If not, maybe remove it or add an assertion.